Docker dev environment (for web + pipeline)

Sybrand opened the pull request:
# Overview
Try it out - let me know what you think. If this works for you. I was going to only do web - but decided to throw in web and pipeline since I didn't hear anything from anyone and had the time.

## Details
- docker-compose.yml file introduced
- package.json:
    - removed flow (not supported on arm64 ubuntu), and script not present in harmony
    - added in mypy - (was not present in package.json)
    - added in translations - (cli-translations not present)
    - modified develop - (removed flow)
- removed a bunch of stuff from scripts that docker compose should be starting up.
- removed instructions to run locally entirely
- removed instructions for installing docker - it was hardware specific, and it keeps changing - just linking to https://docs.docker.com/get-docker/

- @tahitihat please think of the impact re. doing a new cut of harmony given increased divergence of source code.

## Test
Follow instructions in README, in "Local development setup" section.

## Notes about flow
Flow is an issue. The version of flow we currently have is not compatible with arm64 ; arm64 versions are available (if you install the binary manually, not for some reason using npm!) - but we'd have to make a lot of changes outside the scope of this PR to bring the codebase up to what flow is expecting.

jbinary commented:
What is the problem that is solved with this ticket? I mean, I'll need flow and mypy to work anyway, so those should be set up outside of docker thus I still need to set up many things, and what's the matter of having this dockerized thing then?

AbbyBeeler commented on docker-compose.yml:
```diff
@@ -0,0 +1,90 @@
+      postgres:
+        condition: service_healthy
+      hasura:
+        condition: service_started
```
here there's an ordering issue where the first time it's run, the database hasn't been created so hasura fails. I added another comment with a potential approach

Sybrand replied:
I've added in the `./scripts/db/postgres/dev/init_db.py --populate_indicators` to resolve this.

There's no clean answer I guess. Either everything happens automatically, but runs unnecessarily - or it has to be run manually.

AbbyBeeler replied:
I think it's fine to run it manually

AbbyBeeler commented on README.md:
```diff
@@ -93,195 +93,38 @@ Mapbox has a free tier, however there are paid options available. For most users
+### Run Web Server
 
-### Run webpack locally
+1. Start your development environment: `DEFAULT_DRUID_HOST=<public production Druid server IP> ZEN_ENV=<specify environment> docker compose up` (you need to specify druid host and environment)
```
I tried this with `DEFAULT_DRUID_HOST=aws-druid.corp.clambda.com ZEN_ENV=br docker compose up` and it failed with the below error:

```
harmony-web-1       | Traceback (most recent call last):
harmony-web-1       |   File "/app/web/runserver.py", line 97, in <module>
harmony-web-1       |     sys.exit(main())
harmony-web-1       |   File "/app/web/runserver.py", line 92, in main
harmony-web-1       |     app = create_app(flask_config, instance_config, environment)
harmony-web-1       |   File "/app/web/server/app.py", line 455, in create_app
harmony-web-1       |     app = _create_app_internal(flask_config, instance_config, configuration_module)
harmony-web-1       |   File "/app/web/server/app.py", line 186, in _create_app_internal
harmony-web-1       |     _initialize_druid_context(app)
harmony-web-1       |   File "/app/web/server/app.py", line 253, in _initialize_druid_context
harmony-web-1       |     if datasource_config in druid_metadata.get_datasources_for_site(deployment_name):
harmony-web-1       |   File "/app/db/druid/metadata.py", line 98, in get_datasources_for_site
harmony-web-1       |     return [ds.name for ds in self.get_site_datasources() if ds.site == site]
harmony-web-1       |   File "/app/db/druid/metadata.py", line 91, in get_site_datasources
harmony-web-1       |     for ds in self.get_all_datasources()
harmony-web-1       |   File "/app/db/druid/metadata.py", line 81, in get_all_datasources
harmony-web-1       |     r = requests.get(url)
harmony-web-1       |   File "/app/venv/lib/python3.9/site-packages/requests/api.py", line 73, in get
harmony-web-1       |     return request("get", url, params=params, **kwargs)
harmony-web-1       |   File "/app/venv/lib/python3.9/site-packages/requests/api.py", line 59, in request
harmony-web-1       |     return session.request(method=method, url=url, **kwargs)
harmony-web-1       |   File "/app/venv/lib/python3.9/site-packages/requests/sessions.py", line 587, in request
harmony-web-1       |     resp = self.send(prep, **send_kwargs)
harmony-web-1       |   File "/app/venv/lib/python3.9/site-packages/requests/sessions.py", line 695, in send
harmony-web-1       |     adapter = self.get_adapter(url=request.url)
harmony-web-1       |   File "/app/venv/lib/python3.9/site-packages/requests/sessions.py", line 792, in get_adapter
harmony-web-1       |     raise InvalidSchema(f"No connection adapters were found for {url!r}")
harmony-web-1       | requests.exceptions.InvalidSchema: No connection adapters were found for 'aws-druid.corp.clambda.com:8081/druid/coordinator/v1/metadata/datasources'
harmony-web-1       | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
harmony-web-1       | error Command failed with exit code 1.
harmony-web-1 exited with code 1
```

Sybrand replied:
Missing http:// - needs to be `DEFAULT_DRUID_HOST=http://aws-druid.corp.clambda.com ZEN_ENV=br docker compose up` - I've updated the readme with an example.

AbbyBeeler commented on README.md:
```diff
@@ -93,195 +93,38 @@ Mapbox has a free tier, however there are paid options available. For most users
-    - macOS: `psql postgres`
-    - Ubuntu: `psql -U postgres`
-
-5.  Create and upgrade the local postgres database as well as populate the Data Catalog tables: `./scripts/db/postgres/dev/init_db.py --populate_indicators`
```
Adding this step back in could fix the ordering issue. It also handles getting the dimensions populated in Data Catalog, but it's not something that should run every time because it will overwrite all of Data Catalog and read it in from the code.

Sybrand replied:
Added it in as manual step in README

AbbyBeeler commented on Dockerfile_dev:
```diff
@@ -0,0 +1,105 @@
+
+RUN apt-get update \
+    && apt-get -y upgrade \
+    && apt-get -y install --no-install-recommends build-essential \
```
Compared to the previous set up instructions, there are dependencies that aren't here like: jq, pigz, openconnect, watchman, lefthook, coreutils, and grep. I'm not sure why they were added and probably some are because of deployment specific pipeline code, but I think some should be included.

Sybrand replied:
Guilty.

I always question a long list of dependencies - since there's no easy way to validate you actually need them. Was worried about copy/paste creep. I was taking the approach that I'd add them in when things break.

I've added them all in can you please peek in `Dockerfile_dev` and double check that all the dependencies are actually needed?

Sybrand replied:
@AbbyBeeler watchman is being problematic. Having trouble building it. There's no arm .deb ; it builds forever and fails. Seems to be a known issue with watchman right now....

Sybrand replied:
Figured it out - using old watchman from ubuntu, installing python watchman client, and added in a sleep to workaround timeout issues.

AbbyBeeler replied:
Nice. It's used for `yarn translations` btw

AbbyBeeler commented on package.json:
```diff
@@ -127,7 +127,6 @@
-    "develop": "./node_modules/.bin/concurrently --kill-others-on-fail -n flow,webpack,translations,mypy 'yarn flow --continuous_scroll' 'yarn webpack' 'yarn cli-translations watch' 'yarn mypy --continuous_scroll'",
     "flow-check": "echo skipping flow check",
     "init-db": "scripts/db/postgres/dev/init_db.py",
     "postgres": "scripts/db/postgres/dev/connect.py",
```
I wish I could leave comments on other unchanged lines, but `yarn translations` is also a pretty important command to have working since it's used as a part of the integration trainings to create new dimensions. The core script is included, package.json is just missing `"cli-translations": "./scripts/translations/cli_translations.py",`

Sybrand replied:
Hurm. Ok. Looks like them not being there was an oversight?

So might as well try to get them working in this PR I guess.

AbbyBeeler reviewed (changes requested):
Thanks for this Sybrand, it would definitely save a lot of dev set up time! I agree with Sergey that flow and mypy would be necessary before adding this to our repo, but those aren't currently a part of Harmony and so I think this is fine. I ran into a couple issues running the steps that I added comments for, but didn't investigate very thoroughly.

Sybrand commented:
> What is the problem that is solved with this ticket? I mean, I'll need flow and mypy to work anyway, so those should be set up outside of docker thus I still need to set up many things, and what's the matter of having this dockerized thing then?

Harmony doesn't currently include flow or mypy - so I've excluded it.

Super easy to add in mypy - not a problem - if Harmony brings it in, just throw it in.

Flow is a different story. Flow doesn't actually have a distribution for arm on linux as far as I can tell - so it may not actually be a relevant option.

As to what problem is the ticket solving - supporting & documenting how to install and run the application on mac (x86 + arm), ubuntu, wsl is very burdensome.

Sybrand commented on README.md:
```diff
@@ -93,195 +93,38 @@ Mapbox has a free tier, however there are paid options available. For most users
+### Run Pipeline
 
-### Run web server locally
+1. Execute the pipeline `DEFAULT_DRUID_HOST=http://aws-druid.corp.clambda.com COMMAND="./pipeline/br/generate/zeus_generate run" docker compose --profile=pipeline up pipeline`
```
Ok - I see zeus_process is setting `export ZEN_ENV='br'` - so that's fine ; I'm not sure I follow the DATABASE_URL flow exactly, I'll remove `DATABASE_URL` and `ZEN_ENV` from the pipeline config in `docker-compose.yml` and test.

Sybrand replied:
I was missing PYTHONPATH environment variable, so I've added that.

Sybrand commented on docker-compose.yml:
```diff
@@ -0,0 +1,90 @@
+      interval: 10s
+      timeout: 30s
+      retries: 3
+  redis:
```
Ok - I'll remove (looks like it's only used by the celery worker??? will follow up in DM)

jbinary replied:
It's also used for caching (https://github.com/Zenysis/Harmony/blob/4d53cf9071001c372d7ca82086ee5cf8a82c21f6/web/server/security/access_keys.py#L36) but there are different ways for caching.

At Zenysis I unified caching and it uses Redis in prod but falls back to FS caching in Dev, here it's still the older way.

Sybrand replied:
@jbinary should I just bring it back in then? Redis is nice and small - so if it's being use, I don't think it's an issue having it run in docker, standard, as part of dev environment.

Takes dev environment closer to what prod looks like, which is always good.

jbinary replied:
Yeah I think that could be nice but we'd also need to remove those checks for production when we use it.

Could have even more sense if this is ported to harmony: https://phab.zenysis.com/D4056

Sybrand replied:
Are you saying remove the health checks in production?

Sybrand commented on docker/dev/scripts/start_web.sh:
```diff
@@ -0,0 +1,14 @@
+# We need webpack to run in the background.
+source venv/bin/activate && yarn webpack &
+# We need to initialize the database.
+source venv/bin/activate && ZEN_ENV=$ZEN_ENV yarn init-db $ZEN_ENV
```
Doesn't technically need it - yes - I'm just confused around how how dev's generally use it?

`init-db $ZEN_ENV` - the `$ZEN_ENV` feeds into deployments in `init-dp.py` otherwise it does it for all deployments, not just the specified one.

I'll go with your guidance here - you want me to remove it or keep it?

AbbyBeeler replied:
I'd omit init-db here since it doesn't need to run every time the server starts.

It's handy for develop purposes such as if you create a new database migration, then the web server won't start unless you upgrade the database. But it's nice to have to officially upgrade the database if you're working on something incrementally and not have it running automatically. If someone is actively developing against Harmony to the degree they're creating db migrations, then I'd expect them to be comfortable enough with these commands to run `ZEN_ENV=<specify environment> docker compose run web /bin/bash -c "source venv/bin/activate && ./scripts/db/postgres/dev/init_db.py` or `ZEN_ENV=<specify environment> docker compose run web /bin/bash -c "source venv/bin/activate && yarn init-db`.

Sybrand commented on global_config.py:
```diff
@@ -14,7 +15,7 @@
 # Druid is a real time analytics data base. We use it to store all of our
 # analytical data (i.e. data that can be queried in AQT).
-DEFAULT_DRUID_HOST = 'https://druid.hostname.com'
+DEFAULT_DRUID_HOST = os.getenv('DEFAULT_DRUID_HOST', 'https://druid.hostname.com')
```
Going to counter your question with a question - why is there even a 'DEFAULT_DRUID_HOST'? Shouldn't it be removed entirely? How can we assume a default host?

I would think that your configuration is never hard coded, there is no default, and you're always passing in the configuration. So I would expect `global_config.py` not to exist at all.

But yes - valid point - I've removed `https://druid.hostname.com` defaults from `docker-compose.yml` - but would prefer to remover it from global_config too.

AbbyBeeler replied:
I'd agree with not having a fallback because that doesn't make sense and the naming of `DEFAULT_DRUID_HOST` is whack - it should be `DRUID_HOST` (which may be used elsewhere in our project and that's why default was added?). However, I wouldn't want it to have to pass it in every time. I'm not sure if something like that would better live in `global_config.py` or the docker compose, but a project is typically going to have one set Druid. So I'd like to be able to update one place  and then be able to run the web server and without passing in `DEFAULT_DRUID_HOST` every time.

Ideonvn replied:
I agree in that we'd probably want to throw away `global_config.py` entirely in place for something else. Aside from that I'd want us to focus on the real goal here, to get a local docker dev environment up and running. Rather just get it working with all it's annoyances as apposed to trying to fix it all in one go (lets not open that can of worms). This can also just lead to more things to debug when there is an issue.

Sybrand replied:
I'm changing this to `DEFAULT_DRUID_HOST = os.getenv('DEFAULT_DRUID_HOST')`  There should never be hard coded configuration like this in code - that's an anti-pattern.

Sybrand replied:
Actually - couldn't stand seeing DEFAULT_DRUID_HOST - now just DRUID_HOST

Sybrand commented on web/runserver.py:
```diff
@@ -83,23 +75,20 @@ def main():
+            'DATABASE_URL',  # First check for `DATABASE_URL` env
+            Flags.ARGS.db_uri,  # Second check for `--db_uri` arg
+        )
+        or f'postgresql://postgres:@postgres/{db_name}'
```
Gross. So. It was like that, I'll make the change - but got to say - all these defaults peppered all over the place make things pretty difficult. I'm of the opinion we'd be better off dropping that or entirely, and just failing if DATABASE_URL wasn't specified.

AbbyBeeler replied:
Yeah I'd agree, let's just require `DATABASE_URL`

Ideonvn replied:
The main reason for this looks like the ease of use on local vs prod. I had a similar issue on zenysis repo and just fixed the order of preference [here](https://github.com/Zenysis/zenysis/blob/master/web/runserver.py#L86-L92).

Sybrand commented on scripts/db/postgres/dev/upgrade_dev_database.sh:
```diff
@@ -4,17 +4,16 @@ set -o pipefail
 # Test to see if the destination database already exists.
-DB_EXISTS_TEST=$(sudo -u "${POSTGRES_SYSTEM_USER}" \
-  psql postgres -tAc "SELECT 1 FROM pg_database WHERE datname='${DB_NAME}'")
+DB_EXISTS_TEST=$(psql -h ${DATABASE_HOST:-localhost} -U postgres postgres -tAc "SELECT 1 FROM pg_database WHERE datname='${DB_NAME}'")
```
Hmm... as above - why don't I just go for no fallback at all? I'm going to go for `${DATABASE_HOST}` instead of `${DATABASE_HOST:-localhost}`

AbbyBeeler replied:
Yeah a refactor here to be able to just use the DATABASE_URL like `psql DATABASE_URL` would be ideal. It would need to be able to handle the initial case where the table hasn't been created yet so the full DATABASE_URL would fail. I think that would be worth it since `DATABASE_HOST` just exists in this file

Sybrand commented on docker-compose.yml:
```diff
@@ -0,0 +1,90 @@
+      - PYTHONPATH=/app:/usr/bin
+      - DATABASE_HOST=postgres
+      - DATABASE_USER=postgres
+      - DATABASE_URL=postgresql://postgres:zenysis@postgres:5432/${ZEN_ENV}-local
```
I agree - see bigger re-factor notes though. I'll remove here, and rely on defaults in scripts/db/postgres/dev/upgrade_dev_database.sh for now - but really would prefer to remove it entirely. Either only have those, or only have database url!

AbbyBeeler replied:
See other note about removing `DATABASE_HOST`, but yeah just `DATABASE_URL` here would be nice.

Is it possible to construct this `DATABASE_URL` using the environment variables from postgres? Not a big deal at all, but if a user wanted to change their local postgres location, they'd have to update two places in the docker compose.

Sybrand replied:
There's no clean solution to not duplicating the the DATABASE url.

I could move it to a .env file, so it's only specified once - but seems to defeat the purpose, since the entire reason for having docker-compose is so that the postgress location isn't changing.

There's no way to define a DATABASE_URL within the yaml file. I'd love a better solution - but afraid this is it for now.

Sybrand commented on package.json:
```diff
@@ -127,7 +127,6 @@
     "build-clean": "yarn run clean && yarn run build",
     "clean": "rm -rf ./web/public/build/",
-    "develop": "./node_modules/.bin/concurrently --kill-others-on-fail -n flow,webpack,translations,mypy 'yarn flow --continuous_scroll' 'yarn webpack' 'yarn cli-translations watch' 'yarn mypy --continuous_scroll'",
     "flow-check": "echo skipping flow check",
```
I killed the whole line since I assumed it was broken!

- The yarn mypy entry was missing
- flow doesn't run on ubuntu arm
- cli-translations entry was missing

You want me to fix mypy and translations as part of this PR then?

AbbyBeeler replied:
Develop actually is broken lol, nobody uses it because it crashes super frequently and is a nuisance. 

So I would:
- Remove develop
- Add in `cli-translations` - done. My reasoning on including this change in this diff is translations has some dependencies like watchman that are affected by these overall changes so I want to ensure it works.
- Keep `translations` - done
- Remove `flow-check` since it doesn't do anything and won't work with ubuntu
- No opinion on mypy, but seems like you got that working


More broadly and not in the scope of this diff, I think we should add back in the formatting, linting, and flow code that was dropped from our repo. It's annoying to need to manually format code when working in Harmony and lint, flow, and mypy will just help us verify Harmony works and catch errors.

Sybrand commented on scripts/db/postgres/dev/upgrade_dev_database.sh:
```diff
@@ -26,13 +25,12 @@ pushd "${ZEN_SRC_ROOT}" &> /dev/null
 # the extra crap that is needed to actually start the full flask server. We only
 # need the core.
-export DATABASE_URL="postgresql://postgres:@localhost/${DB_NAME}"
+export DATABASE_URL="postgresql://postgres:@${DATABASE_HOST:-localhost}/${DB_NAME}"
```
Yep - see comments above re. trusted - I added it in - but really - like I said before - I think the whole thing needs a re-factor. It's get's very confusing having multiple places with defaults, and multiple ways of constructing DATABASE_URL

Sybrand replied:
I'm hesitant to start doing a re-factor like that as part of this PR though!

AbbyBeeler replied:
Yeah, it's changes that would touch a lot of code and require testing all of those to be sure it works. Also, in order to not make the next Harmony release a pain, we would need to then go update all of those files in our repo.

Ideally what should happen is we don't construct the postgres url anywhere in our code and everything uses `DATABASE_URL`. That currently isn't done because there's no docker compose to set it and I'm not running `DATABASE_URL=postgresql://postgres:zenysis@postgres:5432/br-local yarn br-server` every time I run the web server.

Instead of doing that refactor now, what would you think about changing the values in the docker compose to align with what's hardcoded so it all still works? `POSTGRES_PASSWORD` -> nothing, `POSTGRES_DB` -> `localhost`, and `DATABASE_URL` -> `postgresql://postgres:@localhost:5432/${ZEN_ENV}-local`.

Sybrand commented:
> > What is the problem that is solved with this ticket? I mean, I'll need flow and mypy to work anyway, so those should be set up outside of docker thus I still need to set up many things, and what's the matter of having this dockerized thing then?
> 
> Harmony doesn't currently include flow or mypy - so I've excluded it.
> 
> Super easy to add in mypy - not a problem - if Harmony brings it in, just throw it in.
> 
> Flow is a different story. Flow doesn't actually have a distribution for arm on linux as far as I can tell - so it may not actually be a relevant option.
> 
> As to what problem is the ticket solving - supporting & documenting how to install and run the application on mac (x86 + arm), ubuntu, wsl is very burdensome.

Ok - looks like flow and mypy (and translation!) was an omission from the harmony release cut. I'll quickly try add some of it back in this PR - but flow will fail on M1 - not supported. (which gives me an opportunity to say we should drop flow and switch to typescript...)

AbbyBeeler commented on docker-compose.yml:
```diff
@@ -0,0 +1,90 @@
+      - DATABASE_URL=postgresql://postgres:zenysis@postgres:5432/${ZEN_ENV}-local
+      - ZEN_ENV=${ZEN_ENV}
+      - DEFAULT_DRUID_HOST=${DEFAULT_DRUID_HOST-https://druid.hostname.com}
+      - COMMAND=${COMMAND-./pipeline/br/generate/zeus_generate run}
```
sounds great

Ideonvn commented on util/credentials/provider.py:
```diff
@@ -2,11 +2,13 @@
 
 @contextmanager
 def CredentialProvider(instance_configuration):
+    print(f"going to yield {CREDENTIALS}")
+    print(f'os.getenv("DATABASE_URL") = {os.getenv("DATABASE_URL")}')
```
We'd just want to remove these before merging.

Ideonvn reviewed (commented):
@Sybrand I see some of these files are really just copies from zenysis (necessary I can imagine) & some lint cleanup. Would it make sense to just move those changes into its own PR, making this one smaller and more focused?
- Local DB setup
- scripts/watch/watch_mypy.py
- scripts/watch/watch_util.py

I am sure there is a couple more, I didn't compare zenysis to Harmony.

Ideonvn commented on README.md:
```diff
@@ -49,11 +49,11 @@ After you have written your first data integration (see [writing integrations](#
 ### Introduction
 
-Harmony currently uses [maxbox](https://www.mapbox.com/) for all Map UI visiualisations. In order to use our map visiualisations you would need to have your own `Maxbox Access Token`
+Harmony currently uses [mapbox](https://www.mapbox.com/) for all Map UI visiualisations. In order to use our map visiualisations you would need to have your own `Mapbox Access Token`
```
Whoops 😅

AbbyBeeler commented on README.md:
```diff
@@ -93,195 +93,44 @@ Mapbox has a free tier, however there are paid options available. For most users
+### Run Web Server
+1. Start your development environment: `DEFAULT_DRUID_HOST=http://<public production Druid server IP> ZEN_ENV=<specify environment> docker compose up` e.g. : `DEFAULT_DRUID_HOST=http://aws-druid.corp.clambda.com ZEN_ENV=br docker compose up`
+2. In a seperate terminal, create user account `docker compose exec web /bin/bash -c "source venv/bin/activate && ./scripts/create_user.py --u me@mydomain.com -p admin --first_name=admin --last_name=istrator -a"`
+3. Browse to website on [http://localhost:5000](http://localhost:5000) and log in with the credentials used in step 2.
```
Hasura isn't loading correctly. When I run these steps and then navigate to AQT, there are the following console errors:
```
POST http://localhost:5000/api/graphql [HTTP/1.0 500 INTERNAL SERVER ERROR 387ms]

Warning: RelayResponseNormalizer: Payload did not contain a value for field `dimensionConnection: dimension_connection`. Check that you are parsing with the same query that was used to fetch the payload. [warning.js:30](webpack://zenysis/node_modules/fbjs/lib/warning.js)
Warning: RelayResponseNormalizer: Payload did not contain a value for field `fieldConnection: field_connection`. Check that you are parsing with the same query that was used to fetch the payload. [warning.js:30](webpack://zenysis/node_modules/fbjs/lib/warning.js)
Warning: RelayResponseNormalizer: Payload did not contain a value for field `categoryConnection: category_connection`. Check that you are parsing with the same query that was used to fetch the payload.
```

And when I go to http://localhost:8088/console/ or http://localhost:8080/console/ (the hasura console), there's an error

Sybrand replied:
Strangely I had to specify HASURA_GRAPHQL_ENABLE_CONSOLE: "true" to get the console. Very odd. We don't specify that in our main repo.

Other issue was because of hard coded hasura location. There's a **expletive deleted** piece of code: ```self.HASURA_HOST = getenv(
            'HASURA_HOST',
            global_config.HASURA_HOST if IS_PRODUCTION else 'http://localhost:8088',
        )``` in flask.py.

I've made note of it - prepare for a hotly debated PR in the near future where I rip out all of these defaults. As far as I'm concerned a missing configuration value is a reason for an exception, not a default.

AbbyBeeler commented on requirements-dev.txt:
```diff
@@ -58,3 +58,6 @@ python-jenkins==1.7.0
 factory-boy==3.2.0
+
+# pywatchman gives us watchman-wait which we use for watching translations.
+pywatchman==1.4.1
```
when I run `docker compose exec web /bin/bash -c "source venv/bin/activate && yarn translations"`, I get the following error
```
Starting translations watch server...
Scanning files...
Found 173 files that import I18N
Collecting all translations...
Validating translations...
Updating i18n.js files...
10 files have updated translations
Resolving i18n imports...
Successfully updated all translation files
/bin/sh: 1: watchman-wait: not found
Done in 14.05s.
```

Sybrand replied:
You may just need to do a `docker compose build` - that will probably fix it.

There are some wierdsies around watchman (like it times out the 1st time you run it) - that for the sake of expediency I'm just going to park as is in this PR.

I'd rather just pull it out entirely (given that the code base didn't have this in!) - but seems that translations are pretty key?

AbbyBeeler reviewed (changes requested):
There’s a lot of comments floating around so to summarize the most pressing items to me. These are all items that work in the current Harmony setup flow even if they’re a little time consuming, so it’s important to maintain that:
 - Hasura isn’t running correctly and so the web server has errors
 - I left commentary on which dev dependencies are required. There’s some that are either just useful commands to be able to run or else I couldn’t figure out what they did; no preference on what to do for those. There’s also an issue where pigz is on that list, but not installing correctly and causing errors when the pipeline runs. There’s also an issue where running minio gives an error.
 - Regarding the `DATABASE_URL` being hardcoded all over the place, what do you think about switching to use the values that are hardcoded and so those don’t need to be changed? Then in a later refactor, all of that can be cleaned up in Harmony and our repo? See https://github.com/Zenysis/Harmony/pull/78#discussion_r1110537865
 - There’s an error running yarn translations
 - With the inclusion of yarn translations and mypy, both of those commands show changes or errors: translations updates 10 files and mypy has over 400 errors. The translations changes should be fairly quick to include - just review the update files and double check everything looks alright; the changes appear to be about removed enterprise features. The mypy issues seem to be caused by libraries we import not being typed, so I’m guessing it’s caused by `mypy.ini` not being included in Harmony. However, adding that starts crashing mypy, so I’m not sure if mypy should be split into another PR or stay in here.

And not a critical issue, but most of the provided commands give warnings like 
```
WARN[0000] The "ZEN_ENV" variable is not set. Defaulting to a blank string. 
WARN[0000] The "DEFAULT_DRUID_HOST" variable is not set. Defaulting to a blank string. 
```
even if I'm running the exact provided command, which is a little confusing.

tahitihat commented:
Flagging that I updated the README but intentionally didn't touch the local dev setup section so that merging will be straightforward

Sybrand commented on Dockerfile_dev:
```diff
@@ -0,0 +1,137 @@
+    libproj-dev \
+    libyaml-dev \
+    # @Abby - can you confirm we need everything below this line?
+    dtach \
```
It's a Dockerfile for dev - so there's more leeway to pepper in tools + helpers.

Sybrand commented on Dockerfile_dev:
```diff
@@ -0,0 +1,137 @@
+    pigz \
+    unzip \
+    libopenblas-dev \
+    liblapack-dev \
```
I've added notes around these - I think they may actually be used when compiling python.

Sybrand commented on scripts/data_catalog/populate_query_models_from_config.py:
```diff
@@ -42,14 +42,14 @@
-    'theta_sketch_field',
-    'theta_sketch_size',
-    'type',
+    "children",
```
Stupid black. Black has a preference for ". I'm going to pull it out just to reduce the noise in this PR.

Sybrand commented on Dockerfile_dev:
```diff
@@ -0,0 +1,137 @@
+    "aarch64")\
+    wget https://dl.minio.io/client/mc/release/linux-arm64/mc -O /usr/local/bin/mc;\
+    ;;\
+    esac;
```
Resolved. `chmod +x` didn't work for obscure docker related reasons.

Sybrand commented:
> There’s a lot of comments floating around so to summarize the most pressing items to me. These are all items that work in the current Harmony setup flow even if they’re a little time consuming, so it’s important to maintain that:
> 
> * Hasura isn’t running correctly and so the web server has errors. See [#78 (comment)](https://github.com/Zenysis/Harmony/pull/78#discussion_r1113693205)
> * I left commentary on which dev dependencies are required. There’s some that are either just useful commands to be able to run or else I couldn’t figure out what they did; no preference on what to do for those. There’s also an issue where pigz is on that list, but not installing correctly and causing errors when the pipeline runs (see [#78 (comment)](https://github.com/Zenysis/Harmony/pull/78#discussion_r1110548932)). There’s also an issue where running minio gives an error (see [#78 (comment)](https://github.com/Zenysis/Harmony/pull/78#discussion_r1113688933)).
> * Regarding the `DATABASE_URL` being hardcoded all over the place, what do you think about switching to use the values that are hardcoded and so those don’t need to be changed? Then in a later refactor, all of that can be cleaned up in Harmony and our repo? See [#78 (comment)](https://github.com/Zenysis/Harmony/pull/78#discussion_r1110537865)
> * There’s an error running yarn translations. See [#78 (comment)](https://github.com/Zenysis/Harmony/pull/78#discussion_r1113700253)
> * With the inclusion of yarn translations and mypy, both of those commands show changes or errors: translations updates 10 files and mypy has over 400 errors. The translations changes should be fairly quick to include - just review the update files and double check everything looks alright; the changes appear to be about removed enterprise features. The mypy issues seem to be caused by libraries we import not being typed, so I’m guessing it’s caused by `mypy.ini` not being included in Harmony. However, adding that starts crashing mypy, so I’m not sure if mypy should be split into another PR or stay in here.
> 
> And not a critical issue, but most of the provided commands give warnings like
> 
> ```
> WARN[0000] The "ZEN_ENV" variable is not set. Defaulting to a blank string. 
> WARN[0000] The "DEFAULT_DRUID_HOST" variable is not set. Defaulting to a blank string. 
> ```
> 
> even if I'm running the exact provided command, which is a little confusing.

Re. hasura: Fixed
Re. dev dependencies - I think I've addressed them all - added lots of notes in dockerfile for future reference
Re. DATABASE_URL - I'm just totally ignoring that right now if that's ok - I made a note for a future PR where we just once and for all address it.
Re. Yarn translations - fixed - actual ouput is the job for a future translation
Re. Mypy - I'm just going to pull that out - there's all kinds of complexity introduced here (separate PR!)

AbbyBeeler commented on docker-compose.yml:
```diff
@@ -0,0 +1,106 @@
+      HASURA_GRAPHQL_ENABLE_CONSOLE: "true"
+      HASURA_GRAPHQL_MIGRATIONS_SERVER_TIMEOUT: 500
+      HASURA_GRAPHQL_DATABASE_URL: postgresql://postgres:zenysis@postgres:5432/${ZEN_ENV}-local
+      HASURA_GRAPHQL_METADATA_DIR: '/hasura-metadata/latest'
```
as discussed on slack, removing `HASURA_GRAPHQL_METADATA_DIR` and `HASURA_GRAPHQL_MIGRATIONS_DIR` here as well as `./scripts/db/hasura/dev/container:/zenysis:ro`. and modifying `./graphql/hasura/metadata/versions:/hasura-metadata` -> `./graphql/hasura/metadata/versions/latest:/hasura-metadata`

Sybrand replied:
HASURA_GRAPHQL_METADATA_DIR actually needs to stay
HASURA_GRAPHQL_MIGRATIONS_DIR can be removed - if not present migrations are skipped

./graphql/hasura/metadata/versions:/hasura-metadata is the correct mapping - so not changing that.

AbbyBeeler replied:
It's not particularly important, but to explain my comment more.

The default `HASURA_GRAPHQL_METADATA_DIR` is `/hasura-metadata`.  So I was saying if we change the volume to point `/hasura-metadata` at the latest folder in versions like `./graphql/hasura/metadata/versions/latest:/hasura-metadata`. Then you don't need to additional set the `HASURA_GRAPHQL_METADATA_DIR` to be that latest folder because it's already pointing to it.

Sybrand replied:
Aaah - ok - got it - I'll make that change on https://github.com/Zenysis/Harmony/pull/83

Sybrand commented on .env.example:
```diff
@@ -0,0 +1,2 @@
+DRUID_HOST=http://localhost:8888
```
Changing to `DRUID_HOST=http://druid` to avoid any confusion re. localhost. (localhost would point to the container)

Sybrand commented on Dockerfile_dev:
```diff
@@ -0,0 +1,159 @@
+    # doesn't build on arm64, and they don't provide and arm64 .deb package.
+    # watchman is used by translations to watch files.
+    watchman \
+    vi \
```
vi !== vim - I've added vim

Sybrand replied:
So strange. I swear this worked. Switched to vim.

Sybrand commented:
Introduced docker-compose.yml for local development in a containerised environment, in preference to running web locally.
